site/bps/BP-22-separate-closing-ledgers-from-opening-ledgers.md

```diff
@@ -0,0 +1,97 @@
+
+Then the new API arrived. In the new API, tailing is the primary read use case, and recovery has been demoted to a boolean flag on a builder for the open operation. The user is for the most part unaware of recovery.
+
+However, recovery is still one of the most important aspects of BookKeeper. It is the mechanism on which our Total Order Atomic Broadcast guarantees are built. It deserves to be a bit more prominent than a boolean flag in a builder. It also doesn't help that the terminology is inconsistent. The flag is called withRecovery, while the to check if recovery is needed, we call isClosed. Closed itself is ambiguous because it may refer to the local handle, or it may refer to the state of the ledger.
+
+As tailing is now the primary usecase, we expect that if a writer fails, then whichever node takes over as writer already has a non-recovered ReadHandle open. It would be nice to be able to continue using this Handle to read to the end of the ledger.
+
+### Public Interfaces
+
+I proposes the removal of OpenOpBuilder#withRecovery()
+
+A new method on BookKeeper:
+```
+interface BookKeeper {
+
+    ...
+
+    CompletableFuture<Void> seal(ReadHandle handle);
+}
+```
+
+### Proposed Changes
+
+With the proposed interface changes, every ReadHandle will be opened without recovery. For the tailing usecase, usage will look like.
+
+```java
+ReadHandle reader = bk.newOpenLedgerOp().withLedgerId(X).execute().get();
+long lastReadEntry = -1;
+while (!leader) {
+    long lac = reader.getLastAddConfirmed();
+    if (lac > lastReadEntry) {
+        LedgerEntries entries = reader.read(lastReadEntry+1, lac).get();
+        doSomethingWithEntries(entries);
+        lastReadEntry = lac;
+    }
+}
+assert (leader);
+bk.seal(reader).get();
+long lac = reader.readLastAddConfirmed().get();
+if (lac > lastReadEntry) {
+    LedgerEntries entries = reader.read(lastReadEntry+1, lac).get();
+    doSomethingWithEntries(entries);
+}
+WriteHandle writer = bk.newCreateLedgerOp().execute().get();
+```
+
+Constrast this with how it is with the current recovery on open mechanism.
+
+```
+ReadHandle reader = bk.newOpenLedgerOp().withLedgerId(X).execute().get();
+long lastReadEntry = -1;
+while (!leader) {
+    long lac = reader.getLastAddConfirmed();
+    if (lac > lastReadEntry) {
+        LedgerEntries entries = reader.read(lastReadEntry+1, lac).get();
+        doSomethingWithEntries(entries);
+        lastReadEntry = lac;
+    }
+}
+assert (leader);
+reader.close();
+reader = bk.newOpenLedgerOp().withLedgerId(reader.getId()).withRecovery(true).execute.get();
+long lac = reader.readLastAddConfirmed().get();
+if (lac > lastReadEntry) {
+    LedgerEntries entries = reader.read(lastReadEntry+1, lac).get();
+    doSomethingWithEntries(entries);
+}
+WriteHandle writer = bk.newCreateLedgerOp().execute().get();
+```
+
+The second one is more code, you need to remember to close the previous handle, and the intent of the operation is less clear.
```

Comment:
I am not sure how do you plan to implement `seal(ReadHandle)`. It seems to me it is non-trivial interface thing, you basically need to implement some logic in ReadHandle. I can foresee reusing a readhandle can have a lot of potentially problems. I am preferring the second one (which we currently have now), it has 2-3 lines more code, but it has much clean implementation: when the leadership is changed, the new leader takes over the leadership, open the ledger to fence the writer, and read entries. It deals with ledger metadata changes much better - in other words, I can see the second approach is better for distributedlog implementation. 

frankly speaking, I don't see the approach proposed here is better than existing approach. there is unnecessary code changes introduced and can be error prone. bookkeeper should just provide primitives, rather than dealing with state transition (e.g. follower to leader transition).

Reply:
> I am not sure how do you plan to implement seal(ReadHandle).
In the simplest case it would be trivial. 

```
CompletableFuture<Void> seal(ReadHandle lh) {
    CompletableFuture<Void> future = new CompletableFuture<>();
    this.asyncOpenLedger(lh.getId(), 
        lh.getLedgerMetadata().getPassword(),
        lh.getLedgerMetadata().getDigestType(),
        (rc, lh, ctx) -> {
                if (rc == BKException.Code.OK) {
                    future.complete(null);
                } else {
                    future.completeExceptionally(BKException.create(rc));
                }
         }, null);
    return future;
}
```
The passed in ReadHandle isn't modified at all, it's just used to pull the details of the ledger. I can imagine we would factor out the recovery operation in the future, but right now, all these are still implemented with LedgerHandle.

> It seems to me it is non-trivial interface thing, you basically need to implement some logic in ReadHandle. 

If anything it needs stuff from WriteHandle. Right now we have this really weird class hierarchy. WriteHandle subclasses ReadHandle, but ReadOnlyLedgerHandle subclasses LedgerHandle. Sealing is an writing operation, so it more naturally fits with WriteHandle, but I don't think it even belongs there. It's a one off operation for the ledger, so it belongs at the BookKeeper client.

> It deals with ledger metadata changes much better 
Why do you think it deals with metadata better?

> bookkeeper should just provide primitives, rather than dealing with state transition (e.g. follower to leader transition).
I'm not adding anything about state transitions.

Reply:
> The passed in ReadHandle isn't modified at all, it's just used to pull the details of the ledger. 

you can't simply saying you are not touching ReadHandle. You indeed need to touch ReadHandle, and mutate its state. The propose `seal(ReadHandle)` is effectively same as `ReadHandle#seal()`. I don't like this design. ReadHandle should be non invasive. 

> If anything it needs stuff from WriteHandle. Right now we have this really weird class hierarchy. 

It is not. WriteHandle is effectively a handle that provides write and read, which is use cases like Pulsar needs. ReadHandle is the readonly handle that does non read operations only, which is use cases like DistributedLog, storage interfaces needs. It is as good as what it provides clean interfaces for use cases.

Again, I think this proposal is just another form of #774 . I think `open-to-seal` is more naturally fit in open builder, as a boolean flag (withRecovery(true)), or as a OpenFlag (O_SEAL or O_FENCE or O_RECOVERY or whatever), or whatever name making sense (#sealing() as what @eolivelli proposed). 

As bookkeeper, it should be providing primitives, not dealing with the state transition within a handle.

Reply:
> you can't simply saying you are not touching ReadHandle. You indeed need to touch ReadHandle, and mutate its state. The propose seal(ReadHandle) is effectively same as ReadHandle#seal(). I don't like this design. ReadHandle should be non invasive.

ReadHandle isn't being modified, it's just a container to pull in the detailed required to seal the ledger. If this was C++ it would be const. In fact, it shouldn't be ReadHandle at all, but Handle that is passed in, as that has everything needed in #788 .

> Again, I think this proposal is just another form of #774 . 

In a way, #774 was about changing the default, so that users are forced to take sealing into account. This change takes a different approach, by at least making sealing prominent in the API.

> As bookkeeper, it should be providing primitives, not dealing with the state transition within a handle.

Where am I suggesting we deal with state transitions? All I'm proposing is that we provide an explicit primitive, to allow the transition to take place.

Reply:
btw, i'm fine with leaving it in the builder also to allow "open-to-seal", but it has to be renamed.